### ISSUE TITLE
Learning-rate schedules: optimizer lr is immutable, no warmup/decay support

### ISSUE THREAD

**Issue by michalharakal** [sk-ai-net/skainet#865]
## Summary

`AdamOptimizer` and `SgdOptimizer` take `lr` as a `private val`, so a training loop cannot adjust the learning rate between steps. Standard LLM training recipes (linear warmup + cosine decay, as popularized by GPT-2/3 style pretraining) are therefore impossible without re-creating the optimizer every step — which would discard Adam's moment estimates.

## Proposal

Two small, composable pieces:

1. **Mutable learning rate**: change `lr` to `public var` on `AdamOptimizer` and `SgdOptimizer`. The optimizer reads `lr` in `step()` anyway; making it assignable is enough for any external scheduler.
2. **A schedule primitive**: a `fun interface LrSchedule { fun lrAt(step: Int): Double }` plus a built-in `linearWarmupCosineDecay(totalSteps, warmupSteps, peakLr, initialLr, minLr)` factory — linear ramp from `initialLr` to `peakLr` over the warmup steps, then cosine decay to `minLr`. Training loops then do:

```kotlin
val schedule = linearWarmupCosineDecay(totalSteps, warmupSteps, peakLr = 5e-4)
// each step:
optimizer.lr = schedule.lrAt(globalStep)
trainStep(model, loss, optimizer, ctx, x, y)
```

This keeps the scheduler stateless and decoupled from the optimizer — no wrapper types, no changes to the `Optimizer` interface.

## Context

Needed for porting appendix D (warmup + cosine decay + gradient clipping) of an educational GPT training project to SKaiNET; currently worked around with a project-local AdamW re-implementation whose only difference is `var lr`. PR with the changes above and property tests follows.
